### PR TITLE
Add Sol verified cached client wrapper

### PR DESCRIPTION
**Pull request by vyzaldysanchez**

### CODE REVIEW

File: core/chains/solana/chain.go [hunk 1]
```diff
@@ -44,13 +46,136 @@ type chain struct {
...
+	chainIDVerified     bool
+	chainIDVerifiedLock sync.RWMutex
+
+	solanaclient.ReaderWriter
+}
+
+func (v *verifiedCachedClient) verifyChainID() bool {
+	v.chainIDVerifiedLock.RLock()
+	if v.chainIDVerified {
+		return true
```
Comment (samsondav) [inline core/chains/solana/chain.go:70]: The mutex is never unlocked on this path

File: core/chains/solana/chain.go [hunk 2]
```diff
@@ -44,13 +46,136 @@ type chain struct {
...
+	}
+	v.chainIDVerifiedLock.RUnlock()
+
+	var err error
+
+	v.chainIDVerifiedLock.Lock()
+	defer v.chainIDVerifiedLock.Unlock()
```
Comment (samsondav) [inline core/chains/solana/chain.go:77]: Can a chain ever become unverified? If so you'll need to re-check inside this mutex.

File: core/chains/solana/chain.go [hunk 3]
```diff
@@ -44,13 +46,136 @@ type chain struct {
...
+	v.verifyErr = nil
+	v.chainIDVerified = true
+
+	return v.chainIDVerified
+}
+
+func (v *verifiedCachedClient) SendTx(ctx context.Context, tx *solanago.Transaction) (solanago.Signature, error) {
+	verified := v.verifyChainID()
+	if !verified {
+		return [64]byte{}, v.verifyErr
```
Comment (samsondav) [inline core/chains/solana/chain.go:103]: This can race - access to v.verifyErr is not serialized

**Review by samsondav — CHANGES_REQUESTED** [review]
You got a few races in here

File: core/chains/solana/chain.go (hunk 2 above)
Comment (vyzaldysanchez) [reply core/chains/solana/chain.go:77]: It cannot become unverified, the check only happens on chain creation for the chain id, afterwards it is considered verified, as it was in the previous implementation.

In short, the only status change that will happen is unverified -> verified.

**Review by pinebit — APPROVED** [review]

File: core/chains/solana/chain.go [hunk 4]
```diff
@@ -44,13 +46,133 @@ type chain struct {
...
+	chainIDVerified     bool
+	chainIDVerifiedLock sync.RWMutex
+
+	solanaclient.ReaderWriter
+}
+
+func (v *verifiedCachedClient) verifyChainID() (bool, error) {
+	v.chainIDVerifiedLock.RLock()
+	if v.chainIDVerified {
+		defer v.chainIDVerifiedLock.RUnlock()
```
Comment (samsondav) [inline core/chains/solana/chain.go:69]:
```suggestion
		v.chainIDVerifiedLock.RUnlock()
```

File: core/chains/solana/chain.go [hunk 5]
```diff
@@ -44,13 +46,133 @@ type chain struct {
...
+		return v.chainIDVerified, errors.Errorf("client returned mismatched chain id (expected: %s, got: %s): %s", expectedChainID, v.chainID, v.nodeURL)
+	}
+
+	v.chainIDVerified = true
+
+	return v.chainIDVerified, nil
+}
+
+func (v *verifiedCachedClient) SendTx(ctx context.Context, tx *solanago.Transaction) (solanago.Signature, error) {
+	verified, err := v.verifyChainID()
```
Comment (samsondav) [inline core/chains/solana/chain.go:98]: Would it not make more sense to simply verify the chain once, on `Start`?

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/chains/solana/chain.go (hunk 5 above)
Comment (vyzaldysanchez) [reply core/chains/solana/chain.go:98]: That was the actual issue, that verifying the chain on start was preventing the struct from being created at all. So what we are doing now is to allow the struct to be created and allow for the RPC methods to work if eventually the chain id can be verified.

**Review by samsondav — APPROVED** [review]

**Pull request merged**